File: stack.java
import java.util.Scanner;
import java.util.Stack;

public class stack {
    // Function to return precedence of operators
    static int prec(char c) {
        if (c == '^')
            return 3;
        else if (c == '/' || c == '*')
            return 2;
        else if (c == '+' || c == '-')
            return 1;
        else
            return -1;
    }

    // Function to check if operator is right-associative
    static boolean isRightAssociative(char c) {
        return c == '^';
    }

    public static String infixToPostfix(String s) {
        Stack<Character> st = new Stack<>();
        StringBuilder res = new StringBuilder();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            // If operand, add to result
            if ((c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                (c >= '0' && c <= '9'))
                res.append(c);

            // If '(', push to stack
            else if (c == '(')
                st.push('(');

            // If ')', pop until '( '
            else if (c == ')') {
                while (!st.isEmpty() && st.peek() != '(') {
                    res.append(st.pop());
                }
                st.pop();
            }

            // If operator
            else {
                while (!st.isEmpty() && st.peek() != '(' &&
                       (prec(st.peek()) > prec(c) ||
                       (prec(st.peek()) == prec(c) && !isRightAssociative(c)))) {
                    res.append(st.pop());
                }
                st.push(c);
            }
        }

        // Pop remaining operators
        while (!st.isEmpty()) {
            res.append(st.pop());
        }

        return res.toString();
    }

    static int floorDiv(int a, int b) {
        if (a * b < 0 && a % b != 0)
            return (a / b) - 1;
        return a / b;
    }
    
    public static int evaluatePostfix(String[] arr) {
        Stack<Integer> st = new Stack<>();

        for (String token : arr) {
          
            // If it's an operand (number), push it onto the stack
            if (Character.isDigit(token.charAt(0)) || 
                (token.length() > 1 && token.charAt(0) == '-')) {
                st.push(Integer.parseInt(token));
                System.out.println("Push: " + token + " -> Stack: " + st);
            } 
        
            // Otherwise, it must be an operator
            else {
                int val1 = st.pop();
                int val2 = st.pop();

                System.out.println("Pop: " + val1 + ", " + val2 + " -> Operator: " + token);
                if (token.equals("+")) {
                    int result = val2 + val1;
                    st.push(result);
                    System.out.println(val2 + " + " + val1 + " = " + result);
                    System.out.println("Push: " + result + " -> Stack: " + st);
                }
                else if (token.equals("-")) {
                    int result = val2 - val1;
                    st.push(result);
                    System.out.println(val2 + " - " + val1 + " = " + result);
                    System.out.println("Push: " + result + " -> Stack: " + st);
                }
                else if (token.equals("*")) {
                    int result = val2 * val1;
                    st.push(result);
                    System.out.println(val2 + " * " + val1 + " = " + result);
                    System.out.println("Push: " + result + " -> Stack: " + st);
                }
                else if (token.equals("/")) {
                    int result = floorDiv(val2, val1);
                    st.push(result);
                    System.out.println(val2 + " / " + val1 + " = " + result);
                    System.out.println("Push: " + result + " -> Stack: " + st);
                }
                else if (token.equals("^")) {
                    int result = (int)Math.pow(val2, val1);
                    st.push(result);
                    System.out.println(val2 + " ^ " + val1 + " = " + result);
                    System.out.println("Push: " + result + " -> Stack: " + st);
                }
            }
        }
        return st.pop();
    }
    

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Masukkan ekspresi infix: ");
        String exp = sc.nextLine();

        System.out.println("Ekspresi postfix: " + infixToPostfix(exp));
        String[] arr = infixToPostfix(exp).split("");
        System.out.println("Hasil: " + evaluatePostfix(arr));
        sc.close();
    }
}

File: stack.py
import math

def prec(c):
    if c == '^':
        return 3
    elif c == '/' or c == '*':
        return 2
    elif c == '+' or c == '-':
        return 1
    else:
        return -1

def isRightAssociative(c):
    return c == '^'

def infixToPostfix(s):
    st = []
    res = []

    for c in s:
        # If operand, add to result
        if ('a' <= c <= 'z') or ('A' <= c <= 'Z') or ('0' <= c <= '9'):
            res.append(c)

        # If '(', push to stack
        elif c == '(':
            st.append('(')

        # If ')', pop until '('
        elif c == ')':
            while st and st[-1] != '(':
                res.append(st.pop())
            st.pop()

        # If operator
        else:
            while st and st[-1] != '(' and \
                (prec(st[-1]) > prec(c) or (prec(st[-1]) == prec(c) \
                                    and not isRightAssociative(c))):
                res.append(st.pop())
            st.append(c)

    # Pop remaining operators
    while st:
        res.append(st.pop())

    return ''.join(res)

def evaluatePostfix(arr):
    st = []

    for token in arr:
        
        # If it's an operand (number), push it onto the stack
        if token[0].isdigit() or (len(token) > 1 and token[0] == '-'):
            st.append(int(token))
            print(f"Push: {token} -> Stack: {st}")
        
        # Otherwise, it must be an operator
        else:
            val1 = st.pop()
            val2 = st.pop()

            if token == '+':
                print(f"Pop: {val1}, {val2} -> Operator: {token}")
                st.append(val2 + val1)
                print(f"{val2} + {val1} = {val2 + val1}")
                print(f"Push: {val2 + val1} -> Stack: {st}")
            elif token == '-':
                print(f"Pop: {val1}, {val2} -> Operator: {token}")
                st.append(val2 - val1)
                print(f"{val2} - {val1} = {val2 - val1}")
                print(f"Push: {val2 - val1} -> Stack: {st}")
            elif token == '*':
                print(f"Pop: {val1}, {val2} -> Operator: {token}")
                st.append(val2 * val1)
                print(f"{val2} * {val1} = {val2 * val1}")
                print(f"Push: {val2 * val1} -> Stack: {st}")
            elif token == '/':
                print(f"Pop: {val1}, {val2} -> Operator: {token}")
                st.append(val2 // val1)
                print(f"{val2} // {val1} = {val2 // val1}")
                print(f"Push: {val2 // val1} -> Stack: {st}")
            elif token == '^':
                print(f"Pop: {val1}, {val2} -> Operator: {token}")
                st.append(int(math.pow(val2, val1)))
                print(f"{val2} ^ {val1} = {int(math.pow(val2, val1))}")
                print(f"Push: {int(math.pow(val2, val1))} -> Stack: {st}")
    return st.pop()

if __name__ == '__main__':
    exp = input("Masukkan ekspresi infix: ")
    postfix = infixToPostfix(exp)
    print("Postfix:", postfix)

    arr = list(postfix)
    print("perhitungan Postfix:")
    result = evaluatePostfix(arr)
    print(f"Hasil: {result}")